File: android/app/src/main/java/com/privatepensieve/app/vault/MemoryCardDao.kt
package com.privatepensieve.app.vault

import android.content.ContentValues
import android.database.Cursor
import com.privatepensieve.app.models.MemoryCard
import java.time.Instant
import java.util.UUID

/**
 * Data access object for memory_cards table.
 * All operations local-only. No network.
 */
class MemoryCardDao(private val db: VaultDatabase) {

    // MARK: - Insert

    fun insert(card: MemoryCard) {
        val values = ContentValues().apply {
            put("memory_id", card.memoryId)
            put("created_at", card.createdAt.toString())
            put("updated_at", card.updatedAt.toString())
            put("source_type", "voice")
            put("source_conversation_id", card.sourceConversationId)
            put("title", card.title)
            put("summary", card.summary)
            put("emotion_tags", VaultTypeConverters.toJsonString(card.emotionTags))
            put("topic_tags", VaultTypeConverters.toJsonString(card.topicTags))
            put("people_tags", VaultTypeConverters.toJsonString(card.peopleTags))
            put("place_tags", VaultTypeConverters.toJsonString(card.placeTags))
            put("goal_tags", VaultTypeConverters.toJsonString(card.goalTags))
            put("importance_score", card.importanceScore)
            put("confidence_score", card.confidenceScore)
            put("is_favorite", if (card.isFavorite) 1 else 0)
            put("is_sensitive", if (card.isSensitive) 1 else 0)
            put("is_archived", if (card.isArchived) 1 else 0)
            put("is_deleted", if (card.isDeleted) 1 else 0)
        }
        db.writableDatabase.insertWithOnConflict(
            "memory_cards", null, values,
            android.database.sqlite.SQLiteDatabase.CONFLICT_REPLACE
        )
    }

    // MARK: - Query

    fun fetchAll(): List<MemoryCard> {
        val cursor = db.readableDatabase.rawQuery(
            "SELECT * FROM memory_cards WHERE is_deleted = 0 AND is_archived = 0 ORDER BY created_at DESC",
            null
        )
        return cursor.use { mapCursorToList(it) }
    }

    fun fetchById(id: String): MemoryCard? {
        val cursor = db.readableDatabase.rawQuery(
            "SELECT * FROM memory_cards WHERE memory_id = ?",
            arrayOf(id)
        )
        return cursor.use { if (it.moveToFirst()) mapCursorToCard(it) else null }
    }

    fun fetchFiltered(
        searchText: String? = null,
        tag: String? = null,
        favoritesOnly: Boolean = false,
        minImportance: Int? = null,
        limit: Int? = null
    ): List<MemoryCard> {
        val conditions = mutableListOf("is_deleted = 0", "is_archived = 0")
        val args = mutableListOf<String>()

        if (favoritesOnly) conditions.add("is_favorite = 1")

        minImportance?.let {
            conditions.add("importance_score >= ?")
            args.add(it.toString())
        }

        tag?.let {
            conditions.add("(emotion_tags LIKE ? OR topic_tags LIKE ? OR people_tags LIKE ? OR goal_tags LIKE ?)")
            val wildcard = "%\"$it\"%"
            args.addAll(listOf(wildcard, wildcard, wildcard, wildcard))
        }

        searchText?.let {
            if (it.isNotBlank()) {
                conditions.add("(title LIKE ? OR summary LIKE ?)")
                val wildcard = "%$it%"
                args.addAll(listOf(wildcard, wildcard))
            }
        }

        var sql = "SELECT * FROM memory_cards WHERE ${conditions.joinToString(" AND ")} ORDER BY created_at DESC"
        limit?.let { sql += " LIMIT $it" }

        val cursor = db.readableDatabase.rawQuery(sql, args.toTypedArray())
        return cursor.use { mapCursorToList(it) }
    }

    fun search(query: String, limit: Int = 10): List<MemoryCard> {
        return fetchFiltered(searchText = query, limit = limit)
    }

    // MARK: - Update

    fun toggleFavorite(id: String) {
        db.writableDatabase.execSQL(
            "UPDATE memory_cards SET is_favorite = NOT is_favorite, updated_at = ? WHERE memory_id = ?",
            arrayOf(VaultTypeConverters.nowISO(), id)
        )
    }

    fun updateTags(
        id: String,
        emotionTags: List<String>? = null,
        topicTags: List<String>? = null,
        peopleTags: List<String>? = null,
        goalTags: List<String>? = null
    ) {
        val values = ContentValues().apply {
            put("updated_at", VaultTypeConverters.nowISO())
            emotionTags?.let { put("emotion_tags", VaultTypeConverters.toJsonString(it)) }
            topicTags?.let { put("topic_tags", VaultTypeConverters.toJsonString(it)) }
            peopleTags?.let { put("people_tags", VaultTypeConverters.toJsonString(it)) }
            goalTags?.let { put("goal_tags", VaultTypeConverters.toJsonString(it)) }
        }
        db.writableDatabase.update("memory_cards", values, "memory_id = ?", arrayOf(id))
    }

    // MARK: - Delete

    fun softDelete(id: String) {
        db.writableDatabase.execSQL(
            "UPDATE memory_cards SET is_deleted = 1, updated_at = ? WHERE memory_id = ?",
            arrayOf(VaultTypeConverters.nowISO(), id)
        )
    }

    fun deleteAll() {
        db.writableDatabase.execSQL("DELETE FROM memory_cards")
    }

    // MARK: - Stats

    fun activeCount(): Int {
        val cursor = db.readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM memory_cards WHERE is_deleted = 0", null
        )
        return cursor.use { if (it.moveToFirst()) it.getInt(0) else 0 }
    }

    // MARK: - Cursor Mapping

    private fun mapCursorToList(cursor: Cursor): List<MemoryCard> {
        val list = mutableListOf<MemoryCard>()
        while (cursor.moveToNext()) {
            list.add(mapCursorToCard(cursor))
        }
        return list
    }

    private fun mapCursorToCard(c: Cursor): MemoryCard {
        return MemoryCard(
            memoryId = c.getString(c.getColumnIndexOrThrow("memory_id")),
            createdAt = Instant.parse(c.getString(c.getColumnIndexOrThrow("created_at"))),
            updatedAt = Instant.parse(c.getString(c.getColumnIndexOrThrow("updated_at"))),
            sourceConversationId = c.getString(c.getColumnIndexOrThrow("source_conversation_id")) ?: "",
            title = c.getString(c.getColumnIndexOrThrow("title")),
            summary = c.getString(c.getColumnIndexOrThrow("summary")),
            emotionTags = VaultTypeConverters.fromJsonString(c.getString(c.getColumnIndexOrThrow("emotion_tags"))),
            topicTags = VaultTypeConverters.fromJsonString(c.getString(c.getColumnIndexOrThrow("topic_tags"))),
            peopleTags = VaultTypeConverters.fromJsonString(c.getString(c.getColumnIndexOrThrow("people_tags"))),
            placeTags = VaultTypeConverters.fromJsonString(c.getString(c.getColumnIndexOrThrow("place_tags"))),
            goalTags = VaultTypeConverters.fromJsonString(c.getString(c.getColumnIndexOrThrow("goal_tags"))),
            importanceScore = c.getInt(c.getColumnIndexOrThrow("importance_score")),
            confidenceScore = c.getDouble(c.getColumnIndexOrThrow("confidence_score")),
            isFavorite = c.getInt(c.getColumnIndexOrThrow("is_favorite")) == 1,
            isSensitive = c.getInt(c.getColumnIndexOrThrow("is_sensitive")) == 1,
            isArchived = c.getInt(c.getColumnIndexOrThrow("is_archived")) == 1,
            isDeleted = c.getInt(c.getColumnIndexOrThrow("is_deleted")) == 1
        )
    }
}

File: ios/PrivatePensieve/Vault/MemoryCardDAO.swift
// MemoryCardDAO.swift
// Private Pensieve AI — iOS
// CRUD operations for memory_cards table.
// All data local-only. No cloud. No sync.

import Foundation

/// Data access object for MemoryCard entities.
/// Talks to VaultDatabase directly. No network calls.
final class MemoryCardDAO {

    private let db: VaultDatabase

    init(db: VaultDatabase = .shared) {
        self.db = db
    }

    // MARK: - Insert

    /// Insert a new memory card into the vault.
    func insert(_ card: MemoryCard) throws {
        let sql = """
        INSERT OR REPLACE INTO memory_cards (
            memory_id, created_at, updated_at, source_type, source_conversation_id,
            title, summary, emotion_tags, topic_tags, people_tags, place_tags, goal_tags,
            importance_score, confidence_score, is_favorite, is_sensitive, is_archived, is_deleted
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """
        try db.execute(sql, params: [
            card.id,
            card.createdAt.iso8601String,
            card.updatedAt.iso8601String,
            "voice", // source_type
            card.sourceConversationId,
            card.title,
            card.summary,
            card.emotionTags.jsonString,
            card.topicTags.jsonString,
            card.peopleTags.jsonString,
            card.placeTags.jsonString,
            card.goalTags.jsonString,
            card.importanceScore,
            card.confidenceScore,
            card.isFavorite,
            card.isSensitive,
            card.isArchived,
            card.isDeleted
        ])
    }

    // MARK: - Query

    /// Get all non-deleted, non-archived memory cards, newest first.
    func fetchAll() throws -> [MemoryCard] {
        let sql = """
        SELECT * FROM memory_cards
        WHERE is_deleted = 0 AND is_archived = 0
        ORDER BY created_at DESC
        """
        return try db.query(sql).map { try Self.fromRow($0) }
    }

    /// Get memory cards matching a filter.
    func fetch(filter: MemoryFilter) throws -> [MemoryCard] {
        var sql = "SELECT * FROM memory_cards WHERE is_deleted = 0 AND is_archived = 0"
        var params: [Any?] = []

        if filter.favoritesOnly {
            sql += " AND is_favorite = 1"
        }

        if let minImportance = filter.minImportance {
            sql += " AND importance_score >= ?"
            params.append(minImportance)
        }

        if let tag = filter.tag {
            // Search across all tag columns
            sql += " AND (emotion_tags LIKE ? OR topic_tags LIKE ? OR people_tags LIKE ? OR goal_tags LIKE ?)"
            let wildcard = "%\"\(tag)\"%"
            params.append(contentsOf: [wildcard, wildcard, wildcard, wildcard])
        }

        if let searchText = filter.searchText, !searchText.isEmpty {
            sql += " AND (title LIKE ? OR summary LIKE ?)"
            let wildcard = "%\(searchText)%"
            params.append(contentsOf: [wildcard, wildcard])
        }

        if let daysSince = filter.daysSince {
            let cutoff = Calendar.current.date(byAdding: .day, value: -daysSince, to: Date())!
            sql += " AND created_at >= ?"
            params.append(cutoff.iso8601String)
        }

        sql += " ORDER BY created_at DESC"

        if let limit = filter.limit {
            sql += " LIMIT ?"
            params.append(limit)
        }

        return try db.query(sql, params: params).map { try Self.fromRow($0) }
    }

    /// Get a single memory card by ID.
    func fetchById(_ id: String) throws -> MemoryCard? {
        let sql = "SELECT * FROM memory_cards WHERE memory_id = ?"
        let rows = try db.query(sql, params: [id])
        return try rows.first.map { try Self.fromRow($0) }
    }

    /// Search memory cards by title and summary text.
    func search(query: String, limit: Int = 10) throws -> [MemoryCard] {
        let filter = MemoryFilter(searchText: query, limit: limit)
        return try fetch(filter: filter)
    }

    // MARK: - Update

    /// Toggle favorite status.
    func toggleFavorite(_ id: String) throws {
        let sql = "UPDATE memory_cards SET is_favorite = NOT is_favorite, updated_at = ? WHERE memory_id = ?"
        try db.execute(sql, params: [Date().iso8601String, id])
    }

    /// Update tags on a memory card.
    func updateTags(_ id: String, emotionTags: [String]? = nil, topicTags: [String]? = nil,
                    peopleTags: [String]? = nil, placeTags: [String]? = nil, goalTags: [String]? = nil) throws {
        var setClauses: [String] = ["updated_at = ?"]
        var params: [Any?] = [Date().iso8601String]

        if let tags = emotionTags { setClauses.append("emotion_tags = ?"); params.append(tags.jsonString) }
        if let tags = topicTags { setClauses.append("topic_tags = ?"); params.append(tags.jsonString) }
        if let tags = peopleTags { setClauses.append("people_tags = ?"); params.append(tags.jsonString) }
        if let tags = placeTags { setClauses.append("place_tags = ?"); params.append(tags.jsonString) }
        if let tags = goalTags { setClauses.append("goal_tags = ?"); params.append(tags.jsonString) }

        params.append(id)
        let sql = "UPDATE memory_cards SET \(setClauses.joined(separator: ", ")) WHERE memory_id = ?"
        try db.execute(sql, params: params)
    }

    // MARK: - Delete

    /// Soft-delete a memory card (sets is_deleted = 1).
    func softDelete(_ id: String) throws {
        let sql = "UPDATE memory_cards SET is_deleted = 1, updated_at = ? WHERE memory_id = ?"
        try db.execute(sql, params: [Date().iso8601String, id])
    }

    /// Hard-delete all soft-deleted cards (permanent removal).
    func purgeDeleted() throws {
        try db.execute("DELETE FROM memory_cards WHERE is_deleted = 1")
    }

    /// Delete ALL memory cards (used by Privacy > Delete All Memories).
    func deleteAll() throws {
        try db.execute("DELETE FROM memory_cards")
    }

    // MARK: - Stats

    /// Count of active (non-deleted) memory cards.
    func activeCount() throws -> Int {
        let rows = try db.query("SELECT COUNT(*) as count FROM memory_cards WHERE is_deleted = 0")
        return (rows.first?["count"] as? Int) ?? 0
    }

    // MARK: - Row Mapping

    private static func fromRow(_ row: [String: Any?]) throws -> MemoryCard {
        guard let id = row["memory_id"] as? String,
              let createdStr = row["created_at"] as? String,
              let updatedStr = row["updated_at"] as? String,
              let title = row["title"] as? String,
              let summary = row["summary"] as? String else {
            throw VaultError.executeFailed("Missing required fields in memory_cards row")
        }

        return MemoryCard(
            id: id,
            createdAt: Date.fromISO8601(createdStr) ?? Date(),
            updatedAt: Date.fromISO8601(updatedStr) ?? Date(),
            sourceConversationId: (row["source_conversation_id"] as? String) ?? "",
            title: title,
            summary: summary,
            emotionTags: .fromJSON(row["emotion_tags"] as? String),
            topicTags: .fromJSON(row["topic_tags"] as? String),
            peopleTags: .fromJSON(row["people_tags"] as? String),
            placeTags: .fromJSON(row["place_tags"] as? String),
            goalTags: .fromJSON(row["goal_tags"] as? String),
            importanceScore: (row["importance_score"] as? Int) ?? 5,
            confidenceScore: (row["confidence_score"] as? Double) ?? 0.8,
            duplicateGroupId: nil,
            isFavorite: ((row["is_favorite"] as? Int) ?? 0) == 1,
            isSensitive: ((row["is_sensitive"] as? Int) ?? 0) == 1,
            isArchived: ((row["is_archived"] as? Int) ?? 0) == 1,
            isDeleted: ((row["is_deleted"] as? Int) ?? 0) == 1
        )
    }
}

// MARK: - Filter

/// Filter criteria for querying memory cards.
struct MemoryFilter {
    var searchText: String? = nil
    var tag: String? = nil
    var favoritesOnly: Bool = false
    var minImportance: Int? = nil
    var daysSince: Int? = nil
    var limit: Int? = nil
}
